File: Day19_ComSum.cpp
#include <bits/stdc++.h>
using namespace std;

vector<vector<int>> ans;

void back(vector<int> &can, vector<int> v, int tar)
{
    if (tar == 0)
    {
        ans.push_back(v);
        return;
    }
    if (tar < 0)
    {
        return;
    }
    for (auto &i : can)
    {
        if (v.size() > 0 and v[v.size() - 1] < i)
            continue;
        v.push_back(i);
        back(can, v, tar - i);
        v.pop_back();
    }
}

vector<vector<int>> combinationSum(vector<int> &candidates, int target)
{
    vector<int> v = {};
    back(candidates, v, target);
    return ans;
}

int main()
{

    return 0;
}

File: Day19_ComSum.py
def combinationSum(self, candidates: list[int], target: int) -> list[list[int]]:
    ans = []
    
    def back(can: list[int], l: list[int], tar):
        if(tar == 0):
            ans.append(l)
            return
        if(tar < 0):
            return
        for i in can:
            if(len(l)>0 and l[-1]<i):
                continue
            l.append(i)
            back(can, l, tar-i)
            l.pop(-1)

    l = []
    back(candidates, l, target)
    return ans